fix(tracing): trace frame

musou1500 opened the pull request:
<!--
Thank you for your Pull Request. Please link to the issue this PR addresses.
-->
Fixes: #10845

<!-- PR Checklist
- The issue has been triaged and positive signals were received from maintainers
- Existing tests pass
- New tests are added
- Relevant documentation is changed or added
-->

musou1500 committed:
fix(tracing): trace frame

musou1500 committed:
fix waitUntilPass

musou1500 committed:
fix distilSnapshot to use shortest match

musou1500 committed:
fix snapshot rendering

dgozman commented on packages/playwright-core/src/server/trace/recorder/snapshotterInjected.ts:
```diff
@@ -419,6 +419,8 @@ export function frameSnapshotStreamer(snapshotStreamer: string) {
               continue;
             if (nodeName === 'IFRAME' && name === 'src')
               continue;
+            if (nodeName === 'FRAME' && name === 'src')
```
Let's combine this with previous check like `(nodeName === 'IFRAME' || nodeName === 'FRAME')`. If you rebase on top of the fresh "main" branch, that will make more sense.

dgozman commented on packages/playwright-core/src/web/traceViewer/snapshotRenderer.ts:
```diff
@@ -87,10 +87,10 @@ export class SnapshotRenderer {
+    html = prefix + [
+      `<style>*[__playwright_target__="${this.snapshotName}"] { background-color: #6fa8dc7f; }</style>`,
+      `<script>${snapshotScript()}</script>`
+    ].join('') + html;
```
Why this change?

dgozman replied:
I see, thank you for the explanation. You can now merge hideAllStyle with this `<style>` tag - would be easier to follow.

dgozman commented on packages/playwright-core/src/web/traceViewer/snapshotRenderer.ts:
```diff
@@ -252,7 +251,11 @@ function snapshotScript() {
       document.styleSheets[0].disabled = true;
     };
+
+    const onDOMContentLoaded = () => visit(document);
```
Why this change?

dgozman commented on tests/snapshotter.spec.ts:
```diff
@@ -211,14 +231,32 @@ it.describe('snapshots', () => {
+    let err: unknown;
+    try {
+      await fn(counter);
+    } catch (e) {
```
I don't really like that we rely on the exception - who knows what else could throw in this function? Can we do `captureSnapshotWithFrames` helper that would only do the frames trickery?

musou1500 replied:
Is `captureSnapshotWithFrames` like following?

```typescript
async function captureSnapshotWithFrames(snapshotter: InMemorySnapshotter, page: Page, toImpl:(rpcObject: any) => any) {
  for (let counter = 0; ; ++counter) {
    const snapshot = await snapshotter.captureSnapshot(toImpl(page), 'snapshot' + counter);
    const text = distillSnapshot(snapshot).replace(/frame@[^"]+["]/, '<id>"');
    if (text === '<IFRAME __playwright_src__=\"/snapshot/<id>\"></IFRAME>')
      break;
    await page.waitForTimeout(250);
  }
}
```

dgozman replied:
Yeah, something like that. Or you can just duplicate this code in both tests, that's totally fine and probably easier than doing some helpers.

dgozman reviewed (commented):
Very nice! A few comments below.

musou1500 reviewed (commented):
Thank you for reviewing. I wrote some comments.

dgozman reviewed (commented):
Looks good, let's figure our the tests and simplify a bit.

musou1500 committed:
combine hideAllStyle

musou1500 committed:
remove waitUntilPass

musou1500 commented:
thank you. I fixed.
considering `distilSnapshot` implementation, I added new `hideAllStyle` style tag to array instead of add styles into existing style tag.

dgozman commented:
@musou1500 This looks great, but many bot are unhappy. I suspect that's not your changes. Could you please rebase on top of current `main` branch and push, so I'll rerun the tests.

musou1500 commented:
@dgozman I rebased.

dgozman commented:
Everything looks good, merging.
